src/v/cluster/tests/partition_balancer_planner_test.cc

```diff
@@ -0,0 +1,535 @@
+      = reassignments[0].allocation_units.get_assignments().front().replicas;
+
+    check_expected_assignments(new_replicas_1, expected_nodes);
+    // First move less size node
+    BOOST_REQUIRE_EQUAL(reassignments[0].ntp.tp.topic, "topic-1");
+    BOOST_REQUIRE_EQUAL(reassignments[0].ntp.tp.partition, 2);
+
+    auto new_replicas_2
+      = reassignments[1].allocation_units.get_assignments().front().replicas;
+    check_expected_assignments(new_replicas_2, expected_nodes);
+    BOOST_REQUIRE_EQUAL(reassignments[1].ntp.tp.topic, "topic-1");
+    BOOST_REQUIRE_EQUAL(reassignments[1].ntp.tp.partition, 1);
+}
+
+/*
+ * 5 nodes; 1 topic; 1000 partitions; 1 node down;
+ * Partition size 10_KiB
+ * Batch size 19_MiB
+ * Can move 1900 partitions
+ * Actual
+ *   node_0: partitions: 2000; down: True; disk: unfilled;
+ *   node_1: partitions: 2000; down: False; disk: unfilled;
+ *   node_2: partitions: 2000; down: False; disk: unfilled;
+ *   node_3: partitions: 0; down: False; disk: unfilled;
+ *   node_4: partitions: 0; down: False; disk: unfilled;
+ * Expected
+ *   node_0: partitions: 100;
+ *   node_1: partitions: 2000;
+ *   node_2: partitions: 2000;
+ *   node_3: partitions: 950;
+ *   node_4: partitions: 950;
+ */
+FIXTURE_TEST(test_lot_of_partitions, partition_balancer_planner_fixture) {
+    vlog(logger.debug, "test_lot_of_partitions");
+    allocator_register_nodes(3);
+    create_topic("topic-1", 2000, 3);
+    allocator_register_nodes(2);
+
+    uint64_t local_partition_size = 10_KiB;
+    uint64_t movement_batch_partitions_amount = (reallocation_batch_size
+                                                 + local_partition_size - 1)
+                                                / local_partition_size;
+
+    auto hr = create_health_report(local_partition_size);
+
+    std::set<size_t> unavailable_nodes = {0};
+    auto fm = create_follower_metrics(unavailable_nodes);
+
+    auto reassignments = planner.get_ntp_reassignments(hr, fm);
+    BOOST_REQUIRE_EQUAL(reassignments.size(), movement_batch_partitions_amount);
+
+    std::unordered_set<model::node_id> expected_nodes(
+      {model::node_id(1),
+       model::node_id(2),
+       model::node_id(3),
+       model::node_id(4)});
+
+    size_t node_3_counter = 0;
+    size_t node_4_counter = 0;
+
+    for (auto& reassignment : reassignments) {
+        auto new_replicas
+          = reassignment.allocation_units.get_assignments().front().replicas;
+        BOOST_REQUIRE(new_replicas.size() == 3);
+        for (const auto r : new_replicas) {
+            BOOST_REQUIRE(expected_nodes.contains(r.node_id));
+            if (r.node_id == model::node_id(3)) {
+                node_3_counter += 1;
+            }
+            if (r.node_id == model::node_id(4)) {
+                node_4_counter += 1;
+            }
+        }
+    }
+
+    BOOST_REQUIRE_EQUAL(node_3_counter, node_4_counter);
+    BOOST_REQUIRE_EQUAL(node_4_counter, movement_batch_partitions_amount / 2);
+}
```

Comment:
We should add a scenario: 5 nodes, 3 of them full. Planner should be able to make progress and move partitions away from full nodes.

Reply:
added